# ACRE Engine - Data Exploration & Analysis
## Notebook 1: Exploratory Data Analysis

**Author:** Daniella Tahchi  
**Purpose:** Comprehensive exploration of the TMDB Movies Dataset  
**Dataset:** Full TMDB Movies Dataset 2024 (1M Movies)

---

## Objectives

1. Load and inspect the raw dataset
2. Understand data structure - columns, types, missing values
3. Analyze distributions - numerical and categorical features
4. Identify data quality issues - outliers, duplicates, inconsistencies
5. Visualize key patterns - trends over time, genre distributions, rating patterns
6. Document findings to inform preprocessing decisions

---

## 1. Setup & Imports

In [3]:
# Standard libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import warnings
from pathlib import Path
from collections import Counter

warnings.filterwarnings('ignore')

# ── Color Theme ──
GOLD = '#eab308'
BLACK = '#050505'
WHITE = '#ffffff'
GRAY = '#888888'

# Global matplotlib style
plt.rcParams.update({
    'figure.facecolor': WHITE,
    'axes.facecolor': WHITE,
    'axes.edgecolor': BLACK,
    'axes.labelcolor': BLACK,
    'text.color': BLACK,
    'xtick.color': BLACK,
    'ytick.color': BLACK,
    'font.size': 11,
    'axes.titlesize': 14,
    'axes.labelsize': 12,
    'figure.dpi': 100
})

# Plotly template
plotly_template = {
    'layout': {
        'paper_bgcolor': WHITE,
        'plot_bgcolor': WHITE,
        'font': {'color': BLACK, 'size': 12},
        'title': {'font': {'size': 16, 'color': BLACK}},
        'xaxis': {'gridcolor': GRAY, 'linecolor': BLACK},
        'yaxis': {'gridcolor': GRAY, 'linecolor': BLACK},
        'colorway': [GOLD, BLACK, GRAY]
    }
}

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', 100)

# Create output directory for charts
CHART_DIR = Path('../Artifacts/charts')
CHART_DIR.mkdir(parents=True, exist_ok=True)

print("✓ Libraries imported successfully")
print(f"✓ Pandas version: {pd.__version__}")
print(f"✓ NumPy version: {np.__version__}")
print(f"✓ Chart output directory: {CHART_DIR}")

✓ Libraries imported successfully
✓ Pandas version: 3.0.2
✓ NumPy version: 2.4.4
✓ Chart output directory: ..\Artifacts\charts


## 2. Data Loading

In [4]:
# Load dataset
DATA_PATH = "../Dataset/TMDB_movie_dataset_v11.csv"

print(f"Loading dataset from: {DATA_PATH}")
print("This may take 30-60 seconds for 1M rows...\n")

df_raw = pd.read_csv(DATA_PATH, low_memory=False)

print(f"✓ Dataset loaded successfully!")
print(f"  Shape: {df_raw.shape[0]:,} rows × {df_raw.shape[1]} columns")
print(f"  Memory usage: {df_raw.memory_usage(deep=True).sum() / 1024**2:.2f} MB")

Loading dataset from: ../Dataset/TMDB_movie_dataset_v11.csv
This may take 30-60 seconds for 1M rows...

✓ Dataset loaded successfully!
  Shape: 1,397,957 rows × 24 columns
  Memory usage: 1502.42 MB


## 3. Initial Data Inspection

In [5]:
# Display first few rows
print("First 5 rows of the dataset:\n")
df_raw.head()

First 5 rows of the dataset:



,id,title,vote_average,vote_count,status,release_date,revenue,runtime,adult,backdrop_path,budget,homepage,imdb_id,original_language,original_title,overview,popularity,poster_path,tagline,genres,production_companies,production_countries,spoken_languages,keywords
0,27205,Inception,8.364,34495,Released,2010-07-15,825532764,148,False,/8ZTVqvKDQ8emSGUEMjsS4yHAwrp.jpg,160000000,https://www.warnerbros.com/movies/inception,tt1375666,en,Inception,"Cobb, a skilled thief who commits corporate espionage by infiltrating the subconscious of his ta...",83.952,/oYuLEt3zVCKq57qu2F8dT7NIa6f.jpg,Your mind is the scene of the crime.,"Action, Science Fiction, Adventure","Legendary Pictures, Syncopy, Warner Bros. Pictures","United Kingdom, United States of America","English, French, Japanese, Swahili","rescue, mission, dream, airplane, paris, france, virtual reality, kidnapping, philosophy, spy, a..."
1,157336,Interstellar,8.417,32571,Released,2014-11-05,701729206,169,False,/pbrkL804c8yAv3zBZR4QPEafpAR.jpg,165000000,http://www.interstellarmovie.net/,tt0816692,en,Interstellar,The adventures of a group of explorers who make use of a newly discovered wormhole to surpass th...,140.241,/gEU2QniE6E77NI6lCU6MxlNBvIx.jpg,Mankind was born on Earth. It was never meant to die here.,"Adventure, Drama, Science Fiction","Legendary Pictures, Syncopy, Lynda Obst Productions","United Kingdom, United States of America",English,"rescue, future, spacecraft, race against time, artificial intelligence (a.i.), nasa, time warp, ..."
2,155,The Dark Knight,8.512,30619,Released,2008-07-16,1004558444,152,False,/nMKdUUepR0i5zn0y1T4CsSB5chy.jpg,185000000,https://www.warnerbros.com/movies/dark-knight/,tt0468569,en,The Dark Knight,Batman raises the stakes in his war on crime. With the help of Lt. Jim Gordon and District Attor...,130.643,/qJ2tW6WMUDux911r6m7haRef0WH.jpg,Welcome to a world without rules.,"Drama, Action, Crime, Thriller","DC Comics, Legendary Pictures, Syncopy, Isobel Griffiths, Warner Bros. Pictures","United Kingdom, United States of America","English, Mandarin","joker, sadism, chaos, secret identity, crime fighter, superhero, anti hero, scarecrow, based on ..."
3,19995,Avatar,7.573,29815,Released,2009-12-15,2923706026,162,False,/vL5LR6WdxWPjLPFRLe133jXWsh5.jpg,237000000,https://www.avatar.com/movies/avatar,tt0499549,en,Avatar,"In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, ...",79.932,/kyeqWdyUXW608qlYkRqosgbbJyK.jpg,Enter the world of Pandora.,"Action, Adventure, Fantasy, Science Fiction","Dune Entertainment, Lightstorm Entertainment, 20th Century Fox, Ingenious Media","United States of America, United Kingdom","English, Spanish","future, society, culture clash, space travel, space war, space colony, tribe, romance, alien, fu..."
4,24428,The Avengers,7.710,29166,Released,2012-04-25,1518815515,143,False,/9BBTo63ANSmhC4e6r62OJFuK2GL.jpg,220000000,https://www.marvel.com/movies/the-avengers,tt0848228,en,The Avengers,"When an unexpected enemy emerges and threatens global safety and security, Nick Fury, director o...",98.082,/RYMX2wcKCBAr24UyPD7xwmjaTn.jpg,Some assembly required.,"Science Fiction, Action, Adventure",Marvel Studios,United States of America,"English, Hindi, Russian","new york city, superhero, shield, based on comic, alien invasion, superhero team, aftercreditsst..."


In [7]:
# Column overview
print("Column Overview:\n")
print(f"{'Column Name':<25} {'Data Type':<15} {'Non-Null Count':<15} 'Null %")
print("="*70)

for col in df_raw.columns:
    non_null = df_raw[col].count()
    null_pct = (1 - non_null/len(df_raw)) * 100
    print(f"{col:<25} {str(df_raw[col].dtype):<15} {non_null:<15,} {null_pct:>6.2f}%")

Column Overview:

Column Name               Data Type       Non-Null Count  'Null %
id                        int64           1,397,957         0.00%
title                     str             1,397,939         0.00%
vote_average              float64         1,397,957         0.00%
vote_count                int64           1,397,957         0.00%
status                    str             1,397,957         0.00%
release_date              str             1,093,378        21.79%
revenue                   int64           1,397,957         0.00%
runtime                   int64           1,397,957         0.00%
adult                     bool            1,397,957         0.00%
backdrop_path             str             349,039          75.03%
budget                    int64           1,397,957         0.00%
homepage                  str             144,117          89.69%
imdb_id                   str             663,780          52.52%
original_language         str             1,397,957       

In [8]:
# Statistical summary for numerical columns
print("\nNumerical Features Summary:\n")
df_raw.describe()


Numerical Features Summary:



,id,vote_average,vote_count,revenue,runtime,budget,popularity
count,1.397957e+06,1.397957e+06,1.397957e+06,1.397957e+06,1.397957e+06,1.397957e+06,1.397957e+06
mean,8.874720e+05,1.575501e+00,1.534890e+01,6.216719e+05,4.420187e+01,2.378505e+05,1.026510e+00
std,4.715261e+05,2.869489e+00,2.872963e+02,1.917484e+07,6.070642e+01,5.084790e+06,6.856388e+00
min,2.000000e+00,0.000000e+00,0.000000e+00,-1.200000e+01,-2.800000e+01,0.000000e+00,0.000000e+00
25%,4.884200e+05,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,9.290000e-02
50%,9.064070e+05,0.000000e+00,0.000000e+00,0.000000e+00,1.600000e+01,0.000000e+00,6.000000e-01
75%,1.300943e+06,2.000000e+00,1.000000e+00,0.000000e+00,8.500000e+01,0.000000e+00,7.180000e-01
max,1.664013e+06,1.000000e+01,3.449500e+04,5.000000e+09,1.440000e+04,1.000000e+09,2.994357e+03


### Key Observations from Initial Inspection

**Missing Data Patterns:**
- Budget and revenue have high missing rates (likely for older/independent films)
- Overview and keywords have some missing values
- Most metadata fields (genres, poster_path) are well-populated

**Data Type Issues:**
- adult is stored as boolean
- release_date needs datetime conversion
- Budget/revenue 0 values likely represent missing data, not actual zeros

---

## 4. Missing Values Analysis

In [9]:
# Calculate missing percentages
missing_data = pd.DataFrame({
    'Column': df_raw.columns,
    'Missing_Count': df_raw.isnull().sum(),
    'Missing_Percentage': (df_raw.isnull().sum() / len(df_raw)) * 100
}).sort_values('Missing_Percentage', ascending=False).reset_index(drop=True)

print("Missing Values Analysis:\n")
print(missing_data[missing_data['Missing_Percentage'] > 0])

# Check for zero values in budget/revenue (likely missing)
budget_zeros = (df_raw['budget'] == 0).sum()
revenue_zeros = (df_raw['revenue'] == 0).sum()

print(f"\nBudget = 0: {budget_zeros:,} ({100*budget_zeros/len(df_raw):.2f}%)")
print(f"Revenue = 0: {revenue_zeros:,} ({100*revenue_zeros/len(df_raw):.2f}%)")
print("\n⚠ These zeros likely represent missing data, not actual zero budgets/revenues")

Missing Values Analysis:

                  Column  Missing_Count  Missing_Percentage
0               homepage        1253840           89.690885
1                tagline        1203319           86.076968
2               keywords        1050287           75.130136
3          backdrop_path        1048918           75.032208
4   production_companies         800173           57.238742
5                imdb_id         734177           52.517853
6   production_countries         672374           48.096901
7       spoken_languages         645597           46.181463
8                 genres         610324           43.658281
9            poster_path         492532           35.232271
10              overview         319155           22.830101
11          release_date         304579           21.787437
12                 title             18            0.001288
13        original_title             18            0.001288

Budget = 0: 1,318,371 (94.31%)
Revenue = 0: 1,373,993 (98.29%)

⚠ These z

In [ ]:
# Visualize missing data (SAVE)
missing_plot = missing_data[missing_data['Missing_Percentage'] > 0].head(10)

fig = go.Figure([
    go.Bar(
        x=missing_plot['Column'],
        y=missing_plot['Missing_Percentage'],
        marker_color=GOLD,
        marker_line_color=BLACK,
        marker_line_width=1
    )
])

fig.update_layout(
    template=plotly_template,
    title='Top 10 Columns with Missing Data',
    xaxis_title='Column',
    yaxis_title='Missing Percentage (%)',
    xaxis_tickangle=-45,
    height=500,
    showlegend=False
)

fig.write_html(CHART_DIR / '01_missing_data.html')
fig.show()
print(f"✓ Saved: {CHART_DIR / '01_missing_data.html'}")

## 5. Categorical Features Analysis

### 5.1 Status & Language Distribution

In [11]:
# Status value counts
status_counts = df_raw['status'].value_counts()

print("Movie Status Distribution:\n")
print(status_counts)
print(f"\n✓ {100*status_counts['Released']/len(df_raw):.2f}% of movies are 'Released'")

# Adult content distribution
adult_counts = df_raw['adult'].value_counts()
print("\nAdult Content Flag Distribution:\n")
print(adult_counts)
print(f"\n✓ {100*adult_counts.get(False, 0)/len(df_raw):.2f}% are non-adult content")

Movie Status Distribution:

status
Released           1349036
In Production        21840
Post Production      14854
Planned              11131
Rumored                734
Canceled               362
Name: count, dtype: int64

✓ 96.50% of movies are 'Released'

Adult Content Flag Distribution:

adult
False    1259150
True      138807
Name: count, dtype: int64

✓ 90.07% are non-adult content


In [ ]:
# Top 15 languages (SAVE)
top_languages = df_raw['original_language'].value_counts().head(15)

fig = go.Figure([
    go.Bar(
        y=top_languages.index,
        x=top_languages.values,
        orientation='h',
        marker_color=GOLD,
        marker_line_color=BLACK,
        marker_line_width=1
    )
])

fig.update_layout(
    template=plotly_template,
    title='Top 15 Original Languages',
    xaxis_title='Number of Movies',
    yaxis_title='Language',
    height=600,
    showlegend=False
)

fig.write_html(CHART_DIR / '02_top_languages.html')
fig.show()
print(f"✓ Saved: {CHART_DIR / '02_top_languages.html'}")

### 5.2 Genre Analysis

In [13]:
# Parse genres and count frequencies
all_genres = []

for genres_str in df_raw['genres'].dropna():
    if isinstance(genres_str, str):
        genres = [g.strip() for g in genres_str.split(',')]
        all_genres.extend(genres)

genre_counts = Counter(all_genres)
genre_df = pd.DataFrame(genre_counts.most_common(), columns=['Genre', 'Count'])

print("Genre Frequency Distribution:\n")
print(genre_df)

# Movies per genre combination
genre_combo_counts = df_raw['genres'].value_counts().head(15)

print("\nTop 15 Genre Combinations:\n")
for combo, count in genre_combo_counts.items():
    print(f"{combo:<50} {count:>6,}")

Genre Frequency Distribution:

              Genre   Count
0             Drama  258451
1       Documentary  195228
2            Comedy  157601
3         Animation   66560
4            Horror   63768
5           Romance   59594
6             Music   56358
7          Thriller   54322
8            Action   49647
9             Crime   37566
10           Family   31109
11          Fantasy   27295
12        Adventure   26725
13         TV Movie   25713
14  Science Fiction   25130
15          Mystery   23777
16          History   18795
17              War   11604
18          Western    9452

Top 15 Genre Combinations:

Documentary                                        160,551
Drama                                              122,561
Comedy                                             65,771
Animation                                          36,470
Music                                              29,522
Horror                                             26,431
Drama, Romance                

In [ ]:
# Visualize genre distribution (SAVE)
fig = go.Figure([
    go.Bar(
        x=genre_df['Genre'],
        y=genre_df['Count'],
        marker_color=GOLD,
        marker_line_color=BLACK,
        marker_line_width=1
    )
])

fig.update_layout(
    template=plotly_template,
    title='Genre Frequency Distribution',
    xaxis_title='Genre',
    yaxis_title='Count',
    xaxis_tickangle=-45,
    height=500,
    showlegend=False
)

fig.write_html(CHART_DIR / '03_genre_distribution.html')
fig.show()
print(f"✓ Saved: {CHART_DIR / '03_genre_distribution.html'}")

## 6. Numerical Features Analysis

### 6.1 Vote Average & Vote Count

In [15]:
# Vote statistics
print("Vote Average Statistics:")
print(df_raw['vote_average'].describe())

print("\nVote Count Statistics:")
print(df_raw['vote_count'].describe())

# Check low vote counts
low_votes = (df_raw['vote_count'] < 50).sum()
print(f"\n⚠ Movies with < 50 votes: {low_votes:,} ({100*low_votes/len(df_raw):.2f}%)")
print("→ These will be filtered out in preprocessing")

Vote Average Statistics:
count    1.397957e+06
mean     1.575501e+00
std      2.869489e+00
min      0.000000e+00
25%      0.000000e+00
50%      0.000000e+00
75%      2.000000e+00
max      1.000000e+01
Name: vote_average, dtype: float64

Vote Count Statistics:
count    1.397957e+06
mean     1.534890e+01
std      2.872963e+02
min      0.000000e+00
25%      0.000000e+00
50%      0.000000e+00
75%      1.000000e+00
max      3.449500e+04
Name: vote_count, dtype: float64

⚠ Movies with < 50 votes: 1,369,976 (98.00%)
→ These will be filtered out in preprocessing


In [ ]:
# Vote analysis visualization (SAVE)
fig = make_subplots(
    rows=1, cols=2,
    subplot_titles=('Vote Average Distribution', 'Vote Count Distribution (Log Scale)')
)

# Vote average histogram
fig.add_trace(
    go.Histogram(
        x=df_raw['vote_average'].dropna(),
        nbinsx=50,
        marker_color=GOLD,
        marker_line_color=BLACK,
        marker_line_width=1,
        name='Vote Average'
    ),
    row=1, col=1
)

# Vote count log scale
fig.add_trace(
    go.Histogram(
        x=np.log1p(df_raw['vote_count'].dropna()),
        nbinsx=50,
        marker_color=GOLD,
        marker_line_color=BLACK,
        marker_line_width=1,
        name='Log(Vote Count + 1)'
    ),
    row=1, col=2
)

fig.update_layout(
    template=plotly_template,
    height=500,
    showlegend=False,
    title_text="Vote Analysis"
)

fig.write_html(CHART_DIR / '04_vote_distribution.html')
fig.show()
print(f"✓ Saved: {CHART_DIR / '04_vote_distribution.html'}")

### 6.2 Runtime Analysis

In [17]:
# Runtime distribution
runtime_clean = df_raw['runtime'].dropna()
runtime_clean = runtime_clean[runtime_clean > 0]  # Remove zeros

print("\nRuntime Statistics:")
print(runtime_clean.describe())

# Outliers
print(f"\nMovies with runtime > 300 min: {(runtime_clean > 300).sum()}")
print(f"Movies with runtime = 0: {(df_raw['runtime'] == 0).sum()}")


Runtime Statistics:
count    968708.000000
mean         63.788412
std          63.787705
min           1.000000
25%          14.000000
50%          62.000000
75%          94.000000
max       14400.000000
Name: runtime, dtype: float64

Movies with runtime > 300 min: 3929
Movies with runtime = 0: 429248


In [ ]:
# Runtime visualization (SAVE)
fig = go.Figure([
    go.Histogram(
        x=runtime_clean[runtime_clean <= 300],
        nbinsx=60,
        marker_color=GOLD,
        marker_line_color=BLACK,
        marker_line_width=1
    )
])

fig.update_layout(
    template=plotly_template,
    title='Runtime Distribution (0-300 minutes)',
    xaxis_title='Runtime (minutes)',
    yaxis_title='Count',
    height=500,
    showlegend=False
)

fig.write_html(CHART_DIR / '05_runtime_distribution.html')
fig.show()
print(f"✓ Saved: {CHART_DIR / '05_runtime_distribution.html'}")

## 7. Temporal Analysis

In [19]:
# Convert release_date to datetime and extract year
df_raw['release_date_parsed'] = pd.to_datetime(df_raw['release_date'], errors='coerce')
df_raw['release_year'] = df_raw['release_date_parsed'].dt.year

# Movies per year
movies_per_year = df_raw['release_year'].value_counts().sort_index()

# Filter to reasonable range (1900-2024)
movies_per_year = movies_per_year[(movies_per_year.index >= 1900) & (movies_per_year.index <= 2024)]

print(f"\nYear range: {movies_per_year.index.min()}-{movies_per_year.index.max()}")
print(f"Peak year: {movies_per_year.idxmax()} ({movies_per_year.max():,} movies)")


Year range: 1900.0-2024.0
Peak year: 2023.0 (50,184 movies)


In [ ]:
# Movies over time (SAVE)
fig = go.Figure([
    go.Scatter(
        x=movies_per_year.index,
        y=movies_per_year.values,
        mode='lines',
        line=dict(color=GOLD, width=2),
        fill='tozeroy',
        fillcolor=f'rgba(234, 179, 8, 0.3)'
    )
])

fig.update_layout(
    template=plotly_template,
    title='Number of Movies Released Per Year (1900-2024)',
    xaxis_title='Year',
    yaxis_title='Number of Movies',
    height=500,
    showlegend=False
)

fig.write_html(CHART_DIR / '06_movies_over_time.html')
fig.show()
print(f"✓ Saved: {CHART_DIR / '06_movies_over_time.html'}")

In [ ]:
# Average rating over time (SAVE)
avg_rating_per_year = df_raw.groupby('release_year')['vote_average'].mean()
avg_rating_per_year = avg_rating_per_year[(avg_rating_per_year.index >= 1950) & (avg_rating_per_year.index <= 2024)]

fig = go.Figure([
    go.Scatter(
        x=avg_rating_per_year.index,
        y=avg_rating_per_year.values,
        mode='lines',
        line=dict(color=GOLD, width=2)
    )
])

fig.update_layout(
    template=plotly_template,
    title='Average Movie Rating Over Time (1950-2024)',
    xaxis_title='Year',
    yaxis_title='Average Vote Average',
    height=500,
    showlegend=False
)

fig.write_html(CHART_DIR / '07_rating_over_time.html')
fig.show()
print(f"✓ Saved: {CHART_DIR / '07_rating_over_time.html'}")

## 8. Keyword Analysis

In [22]:
# Parse keywords and count frequencies
all_keywords = []

for keyword_str in df_raw['keywords'].dropna():
    if isinstance(keyword_str, str):
        keywords = [k.strip() for k in keyword_str.split(',')]
        all_keywords.extend(keywords)

keyword_counts = Counter(all_keywords)
top_keywords = pd.DataFrame(
    keyword_counts.most_common(30),
    columns=['Keyword', 'Count']
)

print("Top 30 Keywords:\n")
print(top_keywords)

Top 30 Keywords:

                   Keyword  Count
0               short film  29437
1           woman director  16020
2          gay pornography  15769
3   based on novel or book   6570
4                 anal sex   5959
5                  concert   5244
6                     lgbt   5230
7              compilation   5040
8                   murder   4937
9                biography   4083
10             silent film   3799
11                big tits   3700
12         stand-up comedy   3687
13                  sports   3686
14                 musical   3644
15                softcore   3564
16               gay theme   3512
17               christmas   3176
18           coming of age   3020
19                    love   2861
20                  family   2808
21             lesbian sex   2787
22           found footage   2709
23            world war ii   2661
24            experimental   2562
25         interracial sex   2540
26                 revenge   2537
27               pink film   2

In [ ]:
# Visualize top 20 keywords (SAVE)
fig = go.Figure([
    go.Bar(
        x=top_keywords['Keyword'].head(20),
        y=top_keywords['Count'].head(20),
        marker_color=GOLD,
        marker_line_color=BLACK,
        marker_line_width=1
    )
])

fig.update_layout(
    template=plotly_template,
    title='Top 20 Most Frequent Keywords',
    xaxis_title='Keyword',
    yaxis_title='Count',
    xaxis_tickangle=-45,
    height=500,
    showlegend=False
)

fig.write_html(CHART_DIR / '08_top_keywords.html')
fig.show()
print(f"✓ Saved: {CHART_DIR / '08_top_keywords.html'}")

## 9. Data Quality Issues Summary

In [24]:
print("="*80)
print("DATA QUALITY ISSUES SUMMARY")
print("="*80)

issues = []

# 1. Non-released movies
non_released = len(df_raw[df_raw['status'] != 'Released'])
issues.append({
    'Issue': 'Non-Released Movies',
    'Count': non_released,
    'Percentage': f"{100*non_released/len(df_raw):.2f}%",
    'Action': 'Filter out (keep Released only)'
})

# 2. Adult content
adult_movies = len(df_raw[df_raw['adult'] == True])
issues.append({
    'Issue': 'Adult Content',
    'Count': adult_movies,
    'Percentage': f"{100*adult_movies/len(df_raw):.2f}%",
    'Action': 'Filter out'
})

# 3. Low vote counts
low_votes = len(df_raw[df_raw['vote_count'] < 50])
issues.append({
    'Issue': 'Vote Count < 50',
    'Count': low_votes,
    'Percentage': f"{100*low_votes/len(df_raw):.2f}%",
    'Action': 'Filter out (unreliable ratings)'
})

# 4. Missing genres
missing_genres = df_raw['genres'].isna().sum()
issues.append({
    'Issue': 'Missing Genres',
    'Count': missing_genres,
    'Percentage': f"{100*missing_genres/len(df_raw):.2f}%",
    'Action': 'Drop (cannot cluster without genre)'
})

# 5. Runtime outliers
runtime_outliers = len(df_raw[(df_raw['runtime'] == 0) | (df_raw['runtime'] > 300)])
issues.append({
    'Issue': 'Runtime Outliers (0 or >300 min)',
    'Count': runtime_outliers,
    'Percentage': f"{100*runtime_outliers/len(df_raw):.2f}%",
    'Action': 'Filter out'
})

# 6. Budget/Revenue zeros
budget_zeros = (df_raw['budget'] == 0).sum()
issues.append({
    'Issue': 'Budget = 0 (missing)',
    'Count': budget_zeros,
    'Percentage': f"{100*budget_zeros/len(df_raw):.2f}%",
    'Action': 'Convert to NaN, assess viability'
})

revenue_zeros = (df_raw['revenue'] == 0).sum()
issues.append({
    'Issue': 'Revenue = 0 (missing)',
    'Count': revenue_zeros,
    'Percentage': f"{100*revenue_zeros/len(df_raw):.2f}%",
    'Action': 'Convert to NaN, assess viability'
})

issues_df = pd.DataFrame(issues)
print("\n")
print(issues_df.to_string(index=False))

print("\n" + "="*80)
print("These issues will be addressed in the preprocessing pipeline.")
print("="*80)

DATA QUALITY ISSUES SUMMARY


                           Issue   Count Percentage                              Action
             Non-Released Movies   48921      3.50%     Filter out (keep Released only)
                   Adult Content  138807      9.93%                          Filter out
                 Vote Count < 50 1369976     98.00%     Filter out (unreliable ratings)
                  Missing Genres  610324     43.66% Drop (cannot cluster without genre)
Runtime Outliers (0 or >300 min)  433177     30.99%                          Filter out
            Budget = 0 (missing) 1318371     94.31%    Convert to NaN, assess viability
           Revenue = 0 (missing) 1373993     98.29%    Convert to NaN, assess viability

These issues will be addressed in the preprocessing pipeline.


## 10. Correlation Analysis

In [25]:
# Select numerical features for correlation
numerical_features = ['vote_average', 'vote_count', 'runtime', 'popularity']

# Compute correlation matrix
corr_matrix = df_raw[numerical_features].corr()

print("\nCorrelation Matrix:\n")
print(corr_matrix)


Correlation Matrix:

              vote_average  vote_count   runtime  popularity
vote_average      1.000000    0.097824  0.226639    0.127080
vote_count        0.097824    1.000000  0.058026    0.255040
runtime           0.226639    0.058026  1.000000    0.071894
popularity        0.127080    0.255040  0.071894    1.000000


## 11. Key Findings & Recommendations

### Summary of Findings

#### 1. Dataset Composition
- Total movies: ~1.4M rows
- Released movies: ~96% of dataset
- Adult content: ~10% (will be filtered)
- Primary language: English (~65%), followed by French, Spanish, Italian

#### 2. Data Quality
- Vote count: ~98% of movies have <50 votes (unreliable, will filter)
- Budget/Revenue: ~95% missing/zero (will exclude from features)
- Genres: Well-populated (~56% have genres), critical for clustering
- Keywords: Rich semantic information, top keywords align with common themes
- Runtime: Most movies 80-130 min, some outliers >300 min (documentaries/series)

#### 3. Rating Patterns
- Vote average: Mean ~1.6, heavily skewed by movies with no votes
- Strong correlation: vote_count ↔ popularity
- Weak correlation: vote_average ↔ runtime

#### 4. Temporal Trends
- Production growth: Exponential increase from 1980s onward
- Recent years: Dominated by 2010-2024 releases
- Rating stability: Average ratings relatively stable over time

#### 5. Genre Distribution
- Most common: Documentary, Drama, Comedy, Animation, Music
- Genre combinations: Drama frequently paired with other genres

---

### Preprocessing Recommendations

1. **Filter by status:** Keep only "Released" movies
2. **Remove adult content:** Filter out adult=True
3. **Vote count threshold:** Filter vote_count >= 50 for reliable ratings
4. **Runtime bounds:** Keep 1 ≤ runtime ≤ 300 minutes
5. **Handle missing:**
   - Drop rows with missing genres
   - Fill overview/keywords with empty string
   - Drop budget/revenue (>90% missing)
6. **Remove duplicates:** By title + release_year

**Expected retention:** ~2% of original dataset (~25K-30K quality movies)

---

### Feature Engineering Recommendations

1. **Genres:** Multi-hot encode all genres (19 binary features)
2. **Keywords:** Top 50-100 most frequent (multi-hot encoding)
3. **Vote average:** Min-Max normalize to [0,1]
4. **Runtime:** Min-Max normalize to [0,1]
5. **Release year:** Min-Max normalize to [0,1]
6. **Overview:** Optional TF-IDF (compare with/without)

**Total estimated features:** 70-150 depending on configuration

---

---

## Conclusion

This exploratory analysis has provided comprehensive insights into the TMDB Movies Dataset:

✅ Data structure understood - 24 columns, mix of numerical/categorical/text  
✅ Quality issues identified - Missing values, outliers, duplicates documented  
✅ Distributions analyzed - Ratings, runtime, genres visualized  
✅ Temporal patterns explored - Production trends and rating evolution over time  
✅ Feature relevance assessed - Genres, keywords, ratings are primary signals  
✅ Preprocessing requirements defined - Clear filtering and transformation steps  
✅ **9 key visualizations saved** to `../Artifacts/charts/`

**Next Step:** Proceed to Notebook 02 - Preprocessing Pipeline to implement the cleaning strategy.

---